### PR TITLE
Add ImmutableArray<T>.Builder.DrainToImmutable()

### LINKED ISSUE #72625: [API Proposal]: `ImmutableArray<T>.Builder.ToImmutableAndClear()`
### Background and motivation

In basically every project I use `ImmutableArray` in, I end up creating an extension for `ToImmutableAndClear` on the builder type, which takes advantage of `MoveToImmutable()` if the current builder has right size, and otherwise falls back to `ToImmutable()/Clear()` if not. I would really love it if we added this to the BCL itself.

### API Proposal

```csharp
namespace System.Collections.Immutable;

public struct ImmutableArray<T>
{
    public class Builder
    {
        public ImmutableArray<T> ToImmutableAndClear()
        {
            if (Capacity == Count) return MoveToImmutable();
            else
            {
                var result = ToImmutable();
                Clear();
                return result;
            }
       }
    }
}
```

### API Usage

```csharp
var builder = ImmutableArray.CreateBuilder<string>(/* max possible values */ 42);

foreach (var x in thingGettingMapped)
{
    if (x.Frob) builder.Add(x);
}

return builder.ToImmutableAndClear();
```

### Alternative Designs

_No response_

### Risks

_No response_

### PR DESCRIPTION
**Pull request by jeffhandley**
Closes #72625. Approved API details: https://github.com/dotnet/runtime/issues/72625#issuecomment-1238477596.

`DrainToImmutable` will "do the cheapest thing possible" to return the current contents as an immutable array and then reset the collection to a zero-length array.

### CODE REVIEW

File: src/libraries/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray_1.Builder.cs [hunk 1]
```diff
@@ -213,6 +213,33 @@ public ImmutableArray<T> MoveToImmutable()
...
+                ImmutableArray<T> result;
+
+                if (Capacity == Count)
+                {
+                    result = MoveToImmutable();
+                }
+                else
+                {
+                    result = ToImmutable();
+                    _elements = ImmutableArray<T>.Empty.array!;
```
Comment (333fred) [inline src/libraries/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray_1.Builder.cs:236]: Why not just use `Array<T>.Empty`?

File: src/libraries/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray_1.Builder.cs [hunk 2]
```diff
@@ -213,6 +213,33 @@ public ImmutableArray<T> MoveToImmutable()
...
+                    _count = 0;
+                }
+
+                return result;
```
Comment (stephentoub) [inline src/libraries/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray_1.Builder.cs:240]:
```suggestion
                ImmutableArray<T> result = Capacity == Count ?
                    _elements :
                    ToArray();
                
                _elements = ImmutableArray<T>.Empty.array!;
                _count = 0;

                return result;
```
?

**Review by eiriktsarpalis — APPROVED** [review]
Thanks!

File: src/libraries/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray_1.Builder.cs (hunk 2 above)
Comment (jeffhandley) [reply src/libraries/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray_1.Builder.cs]: Thanks for this coaching, @stephentoub! Refactored per suggestion.

**Review by 333fred — APPROVED** [review]

File: src/libraries/System.Collections.Immutable/tests/ImmutableArrayBuilderTest.cs [hunk 3]
```diff
@@ -880,6 +880,107 @@ public void MoveToImmutableAddRemoveAddToCapacity()
...
+
+        [Fact]
+        public void DrainToImmutablePartialFillClearsCollection()
+        {
+            var builder = CreateBuilderWithCount<int>(4);
+            builder.AddRange(1, 2);
+            builder.DrainToImmutable();
+            builder.Count = 6;
+            Assert.Equal(new[] { 0, 0, 0, 0, 0, 0 }, builder.DrainToImmutable());
+        }
```
Comment (stephentoub) [inline src/libraries/System.Collections.Immutable/tests/ImmutableArrayBuilderTest.cs:982]: May have missed it, but is there a test here that calls DrainToImmutable on an empty builder?

**Review by stephentoub — APPROVED** [review]

File: src/libraries/System.Collections.Immutable/tests/ImmutableArrayBuilderTest.cs (hunk 3 above)
Comment (jeffhandley) [reply src/libraries/System.Collections.Immutable/tests/ImmutableArrayBuilderTest.cs:982]: There were tests that covered that indirectly by draining after clearing or previously draining. I'm adding 2 more tests that will explicitly cover the scenarios of a fully empty builder; thanks for the suggestion.

File: src/libraries/System.Collections.Immutable/tests/ImmutableArrayBuilderTest.cs [hunk 4]
```diff
@@ -880,6 +880,137 @@ public void MoveToImmutableAddRemoveAddToCapacity()
             Assert.Equal(new[] { 0, 1, 2 }, builder.MoveToImmutable());
         }
 
+        [Fact]
+        public void DrainToImmutableEmptyZeroCapacity()
+        {
+            var builder = ImmutableArray.CreateBuilder<int>(0);
```
Comment (bartonjs) [inline src/libraries/System.Collections.Immutable/tests/ImmutableArrayBuilderTest.cs:886]: Literally every use of `var` added in this file in this change is against style guidelines, and should (have) be(en) expanded to the target type.

Comment (stephentoub) [reply src/libraries/System.Collections.Immutable/tests/ImmutableArrayBuilderTest.cs:886]: Yes... but as with Microsoft.Extensions, the whole project started off with a team with different coding guidelines, and we haven't gone through and made them all conform to ours, at which point we've generally opted for new code in the file to conform to the style used by the rest of the file.

Comment (jeffhandley) [reply src/libraries/System.Collections.Immutable/tests/ImmutableArrayBuilderTest.cs:886]: Thanks, and yeah, I did intentionally follow the patterns in the file/project per our [contributing dos and don'ts](https://github.com/dotnet/runtime/blob/main/CONTRIBUTING.md#dos-and-donts).

> **DO** follow our coding style (C# code-specific).
> 
> **DO** give priority to the current style of the project or file you're changing even if it diverges from the general guidelines.

**Pull request merged**